inductiva/localization/localization.py

```diff
@@ -0,0 +1,143 @@
+
+    class _Template(str):
+        """Custom string class to allow for the use of the % operator"""
+
+        def __mod__(self, other):
+            """
+            Allow for the use of the % operator to format the string
+            with the str.format method.
+            """
+            if isinstance(other, dict):
+                return self.format(**other)
+            elif isinstance(other, tuple):
+                return self.format(*other)
+            return self.format(other)
+        
+        def __str__(self) -> str:
+            # hack to allow logging to work with this class.
+            # logging uses msg = str(self.msg) in getMessage() before applying
+            # the % operator. Here we return the object itself to avoid route
+            # the string formatting to the __mod__ method
+            return self
+
+    def __init__(self, lang: Optional[str] = None) -> None:
+        self._translations = {}
+        self._lock = RLock()
+        self._lang = None
+
+        # make sure the default language is loaded
+        # before setting it to the required value
+        self.set_lang(_INDUCTIVA_DEFAULT_LANG)
+        self.set_lang(lang)
+    
+    def get_lang(self):
+        """Get the current language."""
+        return self._lang
+    
+    def set_lang(self, lang: str):
+        """Set the current language.
+        
+        Args:
+            lang (str): Name of the language to set.
+        """
+        with self._lock:
+            # if None is provided, use the system's locale language if
+            # available. Otherwise use the packages's default.
+            if lang is None:
+                _locale, _ = locale.getlocale()
+                lang = _locale.split("_")[0] if _locale else _INDUCTIVA_DEFAULT_LANG
+            
+            if lang in self._translations:
+                return
+
+            self._load(lang)
+            self._lang = lang
+
+    def _load(self, lang: str) -> None:
+        """Load the translation file for the given language.
+
+        NOTE: This method is not thread-safe.
+
+        Args:
+            lang (str): Name of the language to load.
+        """
+        pathname = self.TRANSLATIONS_DIR / f"{lang}.json"
+        
+        if not pathname.exists():
+            # if translation file not available, there is not point on trying
+            # to use a translation for this message. Just use the default.
+            _logger.debug(
+                f"Translation file not available for language '{lang}'. "
+                f"Will use '{_INDUCTIVA_DEFAULT_LANG}' as default language.")
+            return
+
+        with open(pathname, "r") as f:
+            translations = json.load(f)
+
+        for key, translation in translations.items():
+            if isinstance(translation, list):
+                translation = '\n'.join(translation)
+            translations[key] = self._Template(translation)
```

Comment:
Same as above. I think that we can just use `.format()`

Reply:
The problem is to ensure we still provide both old-style `%` and lazy string formatting capabilities so that it works with the `logging` module while giving us some more flexibility
on how we define messages using an external file.

translations might not be necessary now, and they might no even be relevant in the future, but this mechanism gives us the tools to have a centralised location for messages.

In any case, your comment made me think a bit more and I simplified the code. Addressed in [5cb88bc](https://github.com/inductiva/inductiva/pull/981/commits/5cb88bc077487bf855d7d5df6090cb3d9fe952d5).

see also https://docs.python.org/3/howto/logging-cookbook.html#use-of-alternative-formatting-styles